fly/integration/hijack_test.go

```diff
@@ -1045,4 +1047,95 @@ var _ = Describe("Hijacking", func() {
 		})
 	})
+
+	Context("when hijacking yields an executable not found error", func() {
+		var hijacked2 <-chan struct{}
+		JustBeforeEach(func() {
+			didHijack := make(chan struct{})
+			hijacked = didHijack
+			didHijack2 := make(chan struct{})
+			hijacked2 = didHijack2
+			atcServer.AppendHandlers(
+				ghttp.CombineHandlers(
+					ghttp.VerifyRequest("GET", "/api/v1/teams/main/containers", "type=check&resource_name=some-resource-name&pipeline_name=a-pipeline"),
+					ghttp.RespondWithJSONEncoded(200, []atc.Container{
+						{ID: "container-id-1", State: atc.ContainerStateCreated, WorkerName: "some-worker", PipelineName: "a-pipeline", JobName: "", BuildName: "", BuildID: 0, Type: "", StepName: "", ResourceName: "some-resource-name", Attempt: "", User: user},
+					}),
+				),
+				ghttp.CombineHandlers(
+					ghttp.VerifyRequest("GET", "/api/v1/teams/main/containers/container-id-1/hijack"),
+					func(w http.ResponseWriter, r *http.Request) {
+						defer GinkgoRecover()
+
+						conn, err := upgrader.Upgrade(w, r, nil)
+						Expect(err).NotTo(HaveOccurred())
+
+						defer conn.Close()
+
+						close(didHijack)
+
+						var processSpec atc.HijackProcessSpec
+						err = conn.ReadJSON(&processSpec)
+						Expect(err).NotTo(HaveOccurred())
+
+						Expect(processSpec.User).To(Equal(user))
+						Expect(processSpec.Dir).To(Equal(workingDirectory))
+						Expect(processSpec.Path).To(Equal("bash"))
+						Expect(processSpec.Args).To(Equal(args))
+
+						err = conn.WriteJSON(atc.HijackOutput{
+							ExecutableNotFound: true,
+						})
+						Expect(err).NotTo(HaveOccurred())
+
+						err = conn.WriteJSON(atc.HijackOutput{
+							Error: "executable not found",
+						})
+						Expect(err).NotTo(HaveOccurred())
+					},
+				),
+				hijackHandler("container-id-1", didHijack2, nil, "main"),
+			)
+		})
+
+		Context("when a path was not specified", func() {
+			BeforeEach(func() {
+				path = "sh"
+			})
+			It("tries \"bash\" then \"sh\"", func() {
+				os.Stdout.WriteString("\n")
+				flyCmd := exec.Command(flyPath, "-t", targetName, "hijack", "--check", "a-pipeline/some-resource-name")
+
+				stdin, err := flyCmd.StdinPipe()
+				Expect(err).NotTo(HaveOccurred())
+
+				sess, err := gexec.Start(flyCmd, GinkgoWriter, GinkgoWriter)
+				Expect(err).NotTo(HaveOccurred())
+
+				Eventually(hijacked).Should(BeClosed())
+
+				Eventually(sess.Err.Contents).Should(ContainSubstring(ansi.Color("executable not found", "red+b") + "\n"))
+				Eventually(sess.Err.Contents).Should(ContainSubstring("Couldn't find \"bash\" on container, retrying with \"sh\""))
+
+				os.Stdout.WriteString(fmt.Sprintf("ABOUT TO CHECK HIJACKED2: %v\n", hijacked2))
+				Eventually(hijacked2).Should(BeClosed())
+
+				os.Stdout.WriteString("ABOUT TO PUT TO STDIN\n")
+				_, err = fmt.Fprintf(stdin, "some stdin")
+				Expect(err).NotTo(HaveOccurred())
+				_, err = fmt.Fprintf(stdin, "some stdin")
+				Expect(err).NotTo(HaveOccurred())
+
+				err = stdin.Close()
```

Comment:
for some reason this goes through, and the two writes of `"some stdin"` are not?
```
ABOUT TO CLOSE DIDHIJACK: 0xc0006f6060
ABOUT TO READ PROCSPEC
ABOUT TO READ INPUT
ABOUT TO CHECK HIJACKED2: 0xc0006f6060
ABOUT TO PUT TO STDIN
ABOUT TO CHECK STDOUT STDERR
ABOUT TO PARSE INPUT
logging in to team 'main'


target saved
executable not found
Couldn't find "bash" on container, retrying with "sh"

------------------------------
• Failure [0.071 seconds]
Hijacking
/home/shua/src/concourse/fly/integration/hijack_test.go:19
  when hijacking yields an executable not found error
  /home/shua/src/concourse/fly/integration/hijack_test.go:1051
    when a path was not specified
    /home/shua/src/concourse/fly/integration/hijack_test.go:1101
      tries "bash" then "sh" [It]
      /home/shua/src/concourse/fly/integration/hijack_test.go:1105

      Expected
          <atc.HijackInput>: {Closed: true, Stdin: nil, TTYSpec: nil}
      to equal
          <atc.HijackInput>: {Closed: false, Stdin: "some stdin", TTYSpec: nil}

      /home/shua/src/concourse/fly/integration/hijack_test.go:65
```

If I remove this line, then the test fails in the following way. Notice that no `ABOUT TO WRITE STDOUT STDERR` message, which indicates atc side is still waiting on reading input from `conn` even though I've printed a bunch of stuff to `stdin`.
```
ABOUT TO CLOSE DIDHIJACK: 0xc0009b8060
ABOUT TO READ PROCSPEC
ABOUT TO READ INPUT
ABOUT TO CHECK HIJACKED2: 0xc0009b8060
ABOUT TO PUT TO STDIN
ABOUT TO CHECK STDOUT STDERR
logging in to team 'main'


target saved
executable not found
Couldn't find "bash" on container, retrying with "sh"

------------------------------
• Failure [10.072 seconds]
Hijacking
/home/shua/src/concourse/fly/integration/hijack_test.go:19
  when hijacking yields an executable not found error
  /home/shua/src/concourse/fly/integration/hijack_test.go:1051
    when a path was not specified
    /home/shua/src/concourse/fly/integration/hijack_test.go:1101
      tries "bash" then "sh" [It]
      /home/shua/src/concourse/fly/integration/hijack_test.go:1105

      Timed out after 10.000s.
      Got stuck at:
          
      Waiting for:
          some stdout

      /home/shua/src/concourse/fly/integration/hijack_test.go:1130
```

Reply:
@zoetian any pointers for me?

Reply:
Hey @shua, sorry for the delay. @zoetian and I looked into this today, and think we've found the issue.

The `Hijacker` spawns a goroutine that runs `io.Copy` from `pio.In` (stdin of the `fly` process) to a custom `io.Writer` that emits read bytes as `HijackInput` messages.

https://github.com/concourse/concourse/blob/6bdcad882dcb28b9c56a5be16ce31b173804f907/fly/commands/internal/hijacker/hijacker.go#L72-L75

You call `Hijack` twice (once for `bash`, and then once for `sh`), so the goroutine from when we tried to hijack with `bash` is still alive when we start hijacking `sh`. When you write to stdin in the test, it appears to get read in that first goroutine - and since it's already been read, the goroutine for `sh` doesn't see the data.

It'd be nice to cancel the first `io.Copy`, but readers don't support cancelation. We thought of an alternative approach - handle reading stdin outside of the `Hijacker`, feed the result into a `chan atc.HijackInput`, and pass that `chan` into the `Hijacker` in place of `ProcessIO.In`.

You'd also want to pass in a `context.Context` to `Hijacker.Hijack`, and add a `case <-ctx.Done()` to https://github.com/concourse/concourse/blob/6bdcad882dcb28b9c56a5be16ce31b173804f907/fly/commands/internal/hijacker/hijacker.go#L141-L147 (to prevent draining the input chan).

So, `fly/commands/hijack.go` could look something like:

```go
		...
		inputs := make(chan atc.HijackInput, 1)
		go func() {
			var buf [1024]byte
			for {
				n, err := in.Read(buf[:])
				if err != nil {
					break
				}
				inputs <- atc.HijackInput{Stdin: buf[:n]}
			}
			inputs <- atc.HijackInput{Closed: true}
		}()
		io := hijacker.ProcessIO{
			In:  inputs,
			Out: os.Stdout,
			Err: os.Stderr,
		}

		ctx := context.Background()
		h := hijacker.New(target.TLSConfig(), reqGenerator, target.Token())
		result, exeNotFound, err := h.Hijack(ctx, team.Name(), chosenContainer.ID, spec, io)

		if exeNotFound && someShell {
			spec.Path = "sh"
			os.Stderr.WriteString("\rCouldn't find \"bash\" on container, retrying with \"sh\"\n\r")
			result, exeNotFound, err = h.Hijack(ctx, team.Name(), chosenContainer.ID, spec, io)
		}
		return result, err
	}()
	...
```

And `fly/commands/internal/hijacker/hijacker.go` could look like:

```go
type ProcessIO struct {
	In  <-chan atc.HijackInput
	Out io.Writer
	Err io.Writer
}

func (h *Hijacker) Hijack(ctx context.Context, teamName, handle string, spec atc.HijackProcessSpec, pio ProcessIO) (int, bool, error) {
	...
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	go h.monitorTTYSize(ctx, inputs) // handles <-ctx.Done()
	go h.handleInput(ctx, conn, inputs) // handles <-ctx.Done()

	exitStatus, exeNotFound := h.handleOutput(conn, pio)

	return exitStatus, exeNotFound, nil
}
```

---

Let me know if this doesn't make sense/if you have a better approach in mind!

Reply:
thanks for the pointers, I'll try to dig into your suggestions a little and hopefully have an update for this PR sometime this week